### ISSUE TITLE
Token is not active 

### ISSUE THREAD

**Issue by 2jumper3** [openid/AppAuth-iOS#843]
**Describe the bug**
When accessToken trying to refresh I receive error: 
Token is not active 

I'm trying to use this method: 

```
    private func checkTokenAndProceed() async throws -> String {
        try await withCheckedThrowingContinuation { continuation in
            authState?.performAction(freshTokens: { accessToken, _, error in
                if let accessToken {
                    continuation.resume(returning: accessToken)
                    return
                } else {
                    continuation.resume(throwing: error ?? BaseNetworkErrors.badRequest)
                    return
                }
            })
        }
    }
```
But receive this error. First time everything ok, but when token expired I receive this errorЖ 

```
Domain=org.openid.appauth.oauth_token Code=-10 "invalid_grant: Token is not active" UserInfo={OIDOAuthErrorResponseErrorKey={
    error = "invalid_grant";
    "error_description" = "Token is not active";
}, NSLocalizedDescription=invalid_grant: Token is not active, NSUnderlyingError=0x28302f240 {Error Domain=org.openid.appauth.remote-http Code=400 "{"error":"invalid_grant","error_description":"Token is not active"}" UserInfo={NSLocalizedDescription={"error":"invalid_grant","error_description":"Token is not active"}}}}
```

**Comment by ywsang** [comment]
Seems like you're hitting this section of code: [reference](https://github.com/openid/AppAuth-iOS/blob/master/Sources/AppAuthCore/OIDAuthorizationService.m#L472-L490), signaling an issue with the authorization grant ([RFC info](https://datatracker.ietf.org/doc/html/rfc6749#section-5.2)). 

Would you be able to provide more info to assist in debugging? Any info about the status (expiration date, etc.) of the various tokens (refreshToken, lastAuthorizationResponse) could be helpful.

**Comment by 2jumper3** [comment]
Hi @ywsang, and many thanks for your reply! 
After signing in I receive this `OIDAuthState`: 

<OIDAuthState: 0x2830e4a80, isAuthorized: YES, 
refreshToken: "eyJhbG...[redacted]", 
scope: "openid email profile", 
accessToken: "eyJhbG...[redacted]", 
accessTokenExpirationDate: 2024-04-01 08:51:48 +0000, 
idToken: "eyJhbG...[redacted]", 
lastAuthorizationResponse: <OIDAuthorizationResponse: 0x283455900, 
authorizationCode: 3970df1b-6de1-4237-ac54-48d3aa8c9c04.bbc27b-42eb-9f0a-4d6be51ac125.fd8fbbb7-76ec-457b-bad3-3cb33ef9f810, 
state: "hfowzfU-JtAwSyry-niKDleerY4ZTvcRa1XMTsbhI",
 accessToken: "(null)", 
accessTokenExpirationDate: (null), tokenType: (null), idToken: "(null)", scope: "(null)", 
additionalParameters: {
    iss = "https://.....org/realms/client";
    "session_state" = "bbc27767-2cbb-9f0a-4d6be51ac125";
}, 
request: <OIDAuthorizationRequest: 0x283ffbc60, request: https://..../client/protocol/openid-connect/auth?nonce=ZeUrg56l2olJ5Rgee0lt-hjjXDJnxS9f_eXT_XUTgMQ&response_type=code&code_challenge_method=S256&scope=openid%20profile&code_challenge=dlzxDOnvgK86FUQ0sKqyTpq8ayPTr1xFohi8&redirect_uri=/..../hfowzfU-JtAwSyryc8-niKDleerY4Ra1XMTsbhI>>, 
lastTokenResponse: <OIDTokenResponse: 0x2834526c0, 
accessToken: "eyJhbG...[redacted]", 
accessTokenExpirationDate: 2024-04-01 08:51:48 +0000,
 tokenType: Bearer, idToken: "eyJhbG...[redacted]", refreshToken: "eyJhbG...[redacted]", scope: "openid email profile", additionalParameters: {
    "not-before-policy" = 0;
    "refresh_expires_in" = 1799;
    "session_state" = "bbc27767-2cbb-42eb-d6be51ac125";
}, 
request: <OIDTokenRequest: 0x2830bc240, 
request: <URL: https://....../client/protocol/openid-connect/token, 
HTTPBody: code=3970df1b-6de1-4237-ac54-48d3aa8c9c-2cbb-42eb-9f0a-4d6be51ac125.fd8fbbb7-76ec-457b-bad3-3cb33ef9f810&code_verifier=lPhLN7JlzHNgF_mlEMmwsvuCGzjQF5FHRcmioZxaqGQ&redirect_uri=://login&client_id=web&grant_type=authorization_code>>>, lastRegistrationResponse: (null), authorizationError: (null)>

And I suppose that when accessToken expired, and OIDAuthState trying to refresh it with method in the first message, it return error `Token is not active`.

**Comment by mdmathias** [comment]
I'm not exactly sure why this is happening since there are a number of ways this issue can occur. Ultimately, I do not think the issue would be resolved here within AppAuth-iOS. From [section 5.2 of the RFC](https://datatracker.ietf.org/doc/html/rfc6749#section-5.2), `invalid_grant` can refer to any of the following:

```
invalid_grant
    The provided authorization grant (e.g., authorization
    code, resource owner credentials) or refresh token is
    invalid, expired, revoked, does not match the redirection
    URI used in the authorization request, or was issued to
    another client.
```

Looking at how you request and use the token may help you track down this issue. This [blog post](https://blog.timekit.io/google-oauth-invalid-grant-nightmare-and-how-to-fix-it-9f4efaf1da35) may give you some leads to investigate.

I'm closing this for now since I don't think we can resolve this within AppAuth. Please do respond with questions and we will help if possible; or, we will reopen the issue if it is indeed something that can be fixed here.

**Comment by 2jumper3** [comment]
@mdmathias  many thanks for your detail response!
